ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java

```diff
@@ -335,27 +338,46 @@ public static BlockImportResult on_block(
     // Update finalized checkpoint
     final Checkpoint finalizedCheckpoint = state.getFinalized_checkpoint();
     if (finalizedCheckpoint.getEpoch().compareTo(store.getFinalizedCheckpoint().getEpoch()) > 0) {
+      // TODO: Fork Choice Spec does not necessarily ask us to store this checkpoint state, so if we
+      // do error here and exit, we will possibly have acted differently than other nodes.
       try {
         storeCheckpointState(
             store, st, finalizedCheckpoint, store.getBlockState(finalizedCheckpoint.getRoot()));
       } catch (SlotProcessingException | EpochProcessingException e) {
         return BlockImportResult.failedStateTransition(e);
       }
       store.setFinalizedCheckpoint(finalizedCheckpoint);
+      UnsignedLong finalized_slot = store.getFinalizedCheckpoint().getEpochStartSlot();
+      // Update justified if new justified is later than store justified
+      // or if store justified is not in chain with finalized checkpoint
+      if (state
+                  .getCurrent_justified_checkpoint()
+                  .getEpoch()
+                  .compareTo(store.getJustifiedCheckpoint().getEpoch())
+              > 0
+          || !get_ancestor(store, store.getJustifiedCheckpoint().getRoot(), finalized_slot)
+              .equals(store.getFinalizedCheckpoint().getRoot())) {
+        store.setJustifiedCheckpoint(state.getCurrent_justified_checkpoint());
```

Comment:
We need to ensure the checkpoint state for this new justified checkpoint is stored.

Reply:
Done.